doc/developer/design/20240508_subsource_schemas_after_planning.md

```diff
@@ -0,0 +1,113 @@
+# Restructure + genericize subsource planning
+
+-   Associated:
+    -   https://github.com/MaterializeInc/materialize/issues/20208
+    -   https://github.com/MaterializeInc/materialize/pull/26881
+    -   https://github.com/MaterializeInc/materialize/issues/26999
+
+## The Problems
+
+-   To complete #20208, we have to know the schema of the Kafka topic. This
+    schema is most readily knowable after planning the `SOURCE`.
+-   Subsource planning (generating `CreateSubsourceStatement`) currently happens
+    during purification before planning the `SOURCE`.
+-   Purification's current structure also means that we have to implement some
+    bits in purification (for new sources) and some in sequencing (when
+    reconciling additional subsource's definitions with the current source).
+    This would be cleaner if we described the semantics of e.g. source options
+    somewhere like `SourceConnection`.
+
+## Success Criteria
+
+-   We generate `CreateSubourceStatement`s _after_ planning the
+    `CreateSourceStatement`. This unblocks the ability to theoretically create
+    subsources for Kafka's output.
+-   We demonstrate, using PostgreSQL sources as an example, that it is possible
+    to separate "planning" and purification's async concerns, e.g.
+    `PurificationError` wraps some other error that expresses the semantics of
+    handling PostgreSQL `TEXT COLUMNS`.
+
```

Comment:
what about the 'genericizing' of subsource planning? Is that part of the success criteria or just a side-effect of this work?

Reply:
To generate the `CREATE SUBSOURCE` statements after planning the source, I intend to genericize subsource planning.